File: backend/Tickets.Domain/Models/Especificaciones/EspecificacionesImpresora.cs
using System.Text.Json.Serialization;

namespace Tickets.Domain.Models.Especificaciones
{
    /// <summary>
    /// Especificaciones técnicas para Impresoras y Scanners
    /// Se almacena en EspecificacionesJson
    /// </summary>
    public class EspecificacionesImpresora
    {
        [JsonPropertyName("tipoImpresora")]
        public string? TipoImpresora { get; set; } // Laser, Tinta, Multifuncion, Matricial

        [JsonPropertyName("colorOMonocromo")]
        public string? ColorOMonocromo { get; set; } // Color, Monocromo

        [JsonPropertyName("velocidadPPM")]
        public int? VelocidadPPM { get; set; } // Páginas por minuto

        [JsonPropertyName("velocidadPPMColor")]
        public int? VelocidadPPMColor { get; set; }

        [JsonPropertyName("resolucionDPI")]
        public string? ResolucionDPI { get; set; } // 1200x1200, 2400x600

        [JsonPropertyName("cicloDeTrabajo")]
        public int? CicloDeTrabajo { get; set; } // Páginas por mes recomendadas

        [JsonPropertyName("contadorImpresiones")]
        public int? ContadorImpresiones { get; set; }

        // Funciones
        [JsonPropertyName("impresion")]
        public bool? Impresion { get; set; }

        [JsonPropertyName("escaneo")]
        public bool? Escaneo { get; set; }

        [JsonPropertyName("copiado")]
        public bool? Copiado { get; set; }

        [JsonPropertyName("fax")]
        public bool? Fax { get; set; }

        [JsonPropertyName("duplexAutomatico")]
        public bool? DuplexAutomatico { get; set; }

        [JsonPropertyName("alimentadorDocumentos")]
        public bool? AlimentadorDocumentos { get; set; } // ADF

        [JsonPropertyName("capacidadADF")]
        public int? CapacidadADF { get; set; } // Hojas

        // Conectividad
        [JsonPropertyName("wifi")]
        public bool? Wifi { get; set; }

        [JsonPropertyName("ethernet")]
        public bool? Ethernet { get; set; }

        [JsonPropertyName("usb")]
        public bool? USB { get; set; }

        [JsonPropertyName("impresionMovil")]
        public bool? ImpresionMovil { get; set; } // AirPrint, Google Cloud Print

        // Capacidad
        [JsonPropertyName("capacidadBandeja")]
        public int? CapacidadBandeja { get; set; } // Hojas

        [JsonPropertyName("capacidadBandejaSalida")]
        public int? CapacidadBandejaSalida { get; set; }

        [JsonPropertyName("formatosMaximos")]
        public string? FormatosMaximos { get; set; } // A4, Carta, Legal, A3

        // Consumibles
        [JsonPropertyName("modeloTonerTinta")]
        public string? ModeloTonerTinta { get; set; }

        [JsonPropertyName("rendimientoPaginasNegro")]
        public int? RendimientoPaginasNegro { get; set; }

        [JsonPropertyName("rendimientoPaginasColor")]
        public int? RendimientoPaginasColor { get; set; }
    }
}

File: frontend/lib/data/models/equipo/especificaciones/especificaciones_impresora.dart
import 'dart:convert';

/// Especificaciones técnicas para Impresoras y Scanners
class EspecificacionesImpresora {
  final String? tipoImpresora; // Laser, Tinta, Multifuncion, Matricial
  final String? colorOMonocromo; // Color, Monocromo
  final int? velocidadPPM; // Páginas por minuto
  final int? velocidadPPMColor;
  final String? resolucionDPI; // 1200x1200, 2400x600
  final int? cicloDeTrabajo; // Páginas por mes recomendadas
  final int? contadorImpresiones;

  // Funciones
  final bool? impresion;
  final bool? escaneo;
  final bool? copiado;
  final bool? fax;
  final bool? duplexAutomatico;
  final bool? alimentadorDocumentos; // ADF
  final int? capacidadADF; // Hojas

  // Conectividad
  final bool? wifi;
  final bool? ethernet;
  final bool? usb;
  final bool? impresionMovil; // AirPrint, Google Cloud Print

  // Capacidad
  final int? capacidadBandeja; // Hojas
  final int? capacidadBandejaSalida;
  final String? formatosMaximos; // A4, Carta, Legal, A3

  // Consumibles
  final String? modeloTonerTinta;
  final int? rendimientoPaginasNegro;
  final int? rendimientoPaginasColor;

  const EspecificacionesImpresora({
    this.tipoImpresora,
    this.colorOMonocromo,
    this.velocidadPPM,
    this.velocidadPPMColor,
    this.resolucionDPI,
    this.cicloDeTrabajo,
    this.contadorImpresiones,
    this.impresion,
    this.escaneo,
    this.copiado,
    this.fax,
    this.duplexAutomatico,
    this.alimentadorDocumentos,
    this.capacidadADF,
    this.wifi,
    this.ethernet,
    this.usb,
    this.impresionMovil,
    this.capacidadBandeja,
    this.capacidadBandejaSalida,
    this.formatosMaximos,
    this.modeloTonerTinta,
    this.rendimientoPaginasNegro,
    this.rendimientoPaginasColor,
  });

  factory EspecificacionesImpresora.fromJson(Map<String, dynamic> json) {
    return EspecificacionesImpresora(
      tipoImpresora: json['tipoImpresora'],
      colorOMonocromo: json['colorOMonocromo'],
      velocidadPPM: json['velocidadPPM'],
      velocidadPPMColor: json['velocidadPPMColor'],
      resolucionDPI: json['resolucionDPI'],
      cicloDeTrabajo: json['cicloDeTrabajo'],
      contadorImpresiones: json['contadorImpresiones'],
      impresion: json['impresion'],
      escaneo: json['escaneo'],
      copiado: json['copiado'],
      fax: json['fax'],
      duplexAutomatico: json['duplexAutomatico'],
      alimentadorDocumentos: json['alimentadorDocumentos'],
      capacidadADF: json['capacidadADF'],
      wifi: json['wifi'],
      ethernet: json['ethernet'],
      usb: json['usb'],
      impresionMovil: json['impresionMovil'],
      capacidadBandeja: json['capacidadBandeja'],
      capacidadBandejaSalida: json['capacidadBandejaSalida'],
      formatosMaximos: json['formatosMaximos'],
      modeloTonerTinta: json['modeloTonerTinta'],
      rendimientoPaginasNegro: json['rendimientoPaginasNegro'],
      rendimientoPaginasColor: json['rendimientoPaginasColor'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'tipoImpresora': tipoImpresora,
      'colorOMonocromo': colorOMonocromo,
      'velocidadPPM': velocidadPPM,
      'velocidadPPMColor': velocidadPPMColor,
      'resolucionDPI': resolucionDPI,
      'cicloDeTrabajo': cicloDeTrabajo,
      'contadorImpresiones': contadorImpresiones,
      'impresion': impresion,
      'escaneo': escaneo,
      'copiado': copiado,
      'fax': fax,
      'duplexAutomatico': duplexAutomatico,
      'alimentadorDocumentos': alimentadorDocumentos,
      'capacidadADF': capacidadADF,
      'wifi': wifi,
      'ethernet': ethernet,
      'usb': usb,
      'impresionMovil': impresionMovil,
      'capacidadBandeja': capacidadBandeja,
      'capacidadBandejaSalida': capacidadBandejaSalida,
      'formatosMaximos': formatosMaximos,
      'modeloTonerTinta': modeloTonerTinta,
      'rendimientoPaginasNegro': rendimientoPaginasNegro,
      'rendimientoPaginasColor': rendimientoPaginasColor,
    };
  }

  static EspecificacionesImpresora? fromJsonString(String? jsonString) {
    if (jsonString == null || jsonString.isEmpty) return null;
    try {
      return EspecificacionesImpresora.fromJson(json.decode(jsonString));
    } catch (e) {
      return null;
    }
  }

  String toJsonString() {
    return json.encode(toJson());
  }
}
